optuna/samplers/_tpe/parzen_estimator.py

```diff
@@ -18,114 +33,404 @@ class _ParzenEstimatorParameters(
+        weights[np.arange(n_observations), observations] += 1
+        weights /= weights.sum(axis=1, keepdims=True)
+        return weights
+
+    def _calculate_numerical_params(
+        self, observations: np.ndarray, param_name: str
+    ) -> Tuple[np.ndarray, np.ndarray]:
+
+        n_observations = self._n_observations
+        consider_prior = self._parameters.consider_prior
+        consider_endpoints = self._parameters.consider_endpoints
+        consider_magic_clip = self._parameters.consider_magic_clip
+        multivariate = self._parameters.multivariate
+        sigmas0 = self._sigmas0
+        low = self._low[param_name]
+        high = self._high[param_name]
+        assert low is not None
+        assert high is not None
+        assert len(observations) == self._n_observations
+
+        if n_observations == 0:
+            consider_prior = True
+
+        sorted_mus_with_endpoints = np.asarray([], dtype=float)
+        prior_mu = 0.5 * (low + high)
+        prior_sigma = 1.0 * (high - low)
+
         if consider_prior:
-            sorted_weights = numpy.zeros_like(sorted_mus)
-            sorted_weights[:prior_pos] = unsorted_weights[order[:prior_pos]]
-            sorted_weights[prior_pos] = prior_weight
-            sorted_weights[prior_pos + 1 :] = unsorted_weights[order[prior_pos:]]
+            mus = np.empty(n_observations + 1)
+            mus[:n_observations] = observations
+            mus[n_observations] = prior_mu
+            sigmas = np.empty(n_observations + 1)
+        else:
+            mus = observations
+            sigmas = np.empty(n_observations)
+
+        if multivariate:
+            assert sigmas0 is not None
+            sigmas[:] = sigmas0 * (high - low)
         else:
-            sorted_weights = unsorted_weights[order]
-        sorted_weights /= sorted_weights.sum()
+            assert sigmas0 is None
+            sorted_indices = np.argsort(mus)
+            sorted_mus = mus[sorted_indices]
+            sorted_mus_with_endpoints = np.empty(len(mus) + 2, dtype=float)
+            sorted_mus_with_endpoints[0] = low
+            sorted_mus_with_endpoints[1:-1] = sorted_mus
+            sorted_mus_with_endpoints[-1] = high
+
+            sorted_sigmas = np.maximum(
+                sorted_mus_with_endpoints[1:-1] - sorted_mus_with_endpoints[0:-2],
+                sorted_mus_with_endpoints[2:] - sorted_mus_with_endpoints[1:-1],
+            )
+            sigmas[:] = sorted_sigmas[np.argsort(sorted_indices)]
 
-        # We adjust the range of the 'sigma' according to the 'consider_magic_clip' flag.
+        if consider_prior:
+            sigmas[n_observations] = prior_sigma
+
+        if not multivariate and not consider_endpoints and sorted_mus_with_endpoints.shape[0] >= 4:
+            sigmas[np.argmin(mus)] = sorted_mus_with_endpoints[2] - sorted_mus_with_endpoints[1]
+            sigmas[np.argmax(mus)] = sorted_mus_with_endpoints[-2] - sorted_mus_with_endpoints[-3]
```

Comment:
Probably better (faster and less error prone) to do this right before unsorting `sigmas` as line 397 (`sigmas[:] = sorted_sigmas[np.argsort(sorted_indices)]`) since this code is executed with the default TPE configuration (`consider_endopints=False`).